rmf_task_sequence/src/rmf_task_sequence/events/PerformAction.cpp

```diff
@@ -0,0 +1,272 @@
+{
+  return _invariant_finish_state;
+}
+
+//==============================================================================
+class PerformAction::Description::Implementation
+{
+public:
+  std::string category;
+  nlohmann::json description;
+  rmf_traffic::Duration action_duration_estimate;
+  bool use_tool_sink;
+  std::optional<Location> expected_finish_location;
+};
+
+//==============================================================================
+auto PerformAction::Description::make(
+  const std::string& category,
+  nlohmann::json action,
+  rmf_traffic::Duration duration,
+  bool use_tool_sink,
+  std::optional<PerformAction::Location> location) -> DescriptionPtr
+{
+  auto description = std::shared_ptr<Description>(new Description);
+  description->_pimpl = rmf_utils::make_impl<Implementation>(
+    Implementation
+    {
+      std::move(category),
+      std::move(action),
+      std::move(duration),
+      std::move(use_tool_sink),
+      std::move(location)
+    });
+
+  return description;
+}
+
+//==============================================================================
+PerformAction::Description::Description()
+{
+  // Do nothing
+}
+
+//==============================================================================
+const std::string&
+PerformAction::Description::category() const
+{
+  return _pimpl->category;
+}
+
+//==============================================================================
+auto PerformAction::Description::category(
+  const std::string& new_category) -> Description&
+{
+  _pimpl->category = new_category;
+  return *this;
+}
+
+//==============================================================================
+const nlohmann::json&
+PerformAction::Description::description() const
+{
+  return _pimpl->description;
+}
+
+//==============================================================================
+auto PerformAction::Description::description(
+  const nlohmann::json& new_description) -> Description&
+{
+  _pimpl->description = new_description;
+  return *this;
+}
+
+//==============================================================================
+const rmf_traffic::Duration&
+PerformAction::Description::action_duration_estimate() const
+{
+  return _pimpl->action_duration_estimate;
+}
+
+//==============================================================================
+auto PerformAction::Description::action_duration_estimate(
+  rmf_traffic::Duration new_duration) -> Description&
+{
+  _pimpl->action_duration_estimate = std::move(new_duration);
+  return *this;
+}
+
+//==============================================================================
+bool PerformAction::Description::use_tool_sink() const
+{
+  return _pimpl->use_tool_sink;
+}
+
+//==============================================================================
+auto PerformAction::Description::use_tool_sink(
+  bool use_tool) -> Description&
+{
+  _pimpl->use_tool_sink = std::move(use_tool);
```

Comment:
Not really important, but just a remark: Moving a primitive type like `bool` doesn't really do anything special compared to copying it. Either way it will perform a copy.

Moving only matters when one or more fields are heap-allocated objects whose copy operators perform a deep copy, like `std::vector` or `std::[unordered_]map`. Or objects whose copy operator has side effects, like `std::shared_ptr`'s copy operation increments a reference count while its move operation doesn't. Or most importantly objects that don't offer a copy operator like `std::unique_ptr`.

Using move here doesn't hurt anything, but I thought it would be good to share the above info.

Reply:
Thanks for the explanation; very helpful 👍🏼